### DISCUSSION TITLE
Fitting with Pearson4 shape

### DISCUSSION THREAD

**Discussion by nicocopez** [h2020charisma/ramanchada2#106] [Q&A] [answered]
I remember there were some discussions around the peak fitting with a Pearson4 shape. However, for what i see, it was not implemented, or it is not working properly. Am I missing something and therefore it is not working? Or shouldn't work yet? Is there any plan to implement it?
Thank you,

**Comment by vedina** [comment, marked as answer]
it has been implemented long ago https://github.com/h2020charisma/ramanchada2/blob/1bf736e04c152bdee7a4fc92d56be9837792cead/src/ramanchada2/spectrum/peaks/fit_peaks.py#L19

**Reply by nicocopez** [reply]
Thank you for the quick answer. It seems there is some issue, as it does not work as with the other shapes. For example:
![imatge](https://github.com/h2020charisma/ramanchada2/assets/47475896/dec34913-ff8d-4d71-a0f1-0392c74de1fb)
![imatge](https://github.com/h2020charisma/ramanchada2/assets/47475896/9e64ac82-48c0-48a3-a093-768042619849)
![imatge](https://github.com/h2020charisma/ramanchada2/assets/47475896/aa157cd7-6b1d-461e-8cdd-e3e342b03739)
I assume I am missing something, but was not able to find what. Any clue?

**Comment by vedina** [comment]
Please provide test case

**Reply by nicocopez** [reply]
Here I attach the x and y axis values and the code I am using. It is working for Pearson7 for example, but not for pearson4
[x_axis.txt](https://github.com/h2020charisma/ramanchada2/files/14589691/x_axis.txt)
[y_axis.txt](https://github.com/h2020charisma/ramanchada2/files/14589693/y_axis.txt)

```
x1 = np.loadtxt('x_axis.txt')
y1 = np.loadtxt('y_axis.txt')
spe0 = rc2.spectrum.Spectrum(x=x1, y=y1)

peak_candidates = spe0.find_peak_multipeak(sharpening= None, strategy='topo', hht_chain = [150], wlen = 200)
fitres = spe0.fit_peak_multimodel(profile='Pearson7', candidates=peak_candidates, no_fit=False)

fig, ax = plt.subplots(figsize=(35, 10))
fitres.plot(ax=ax, individual_peaks=False)
spe0.plot(ax=ax)
```

**Reply by vedina** [reply]
@evg-mar may be you could check this issue as part of learning ramanchada2

**Reply by evg-mar** [reply]
@georgievgeorgi 
Could you please take a look at this changes in Pearson4. Could you confirm that it is correct? And also where did you get the formulas for Pearson4, 7, Voigt,.. etc. Was it from any official documentation of the project or from some book / wikipedia, etc.

I changed _height to _amplitude in ->
fit_params[f'p{peak_i}_height'].set(value=peak.amplitude) to
fit_params[f'p{peak_i}_amplitude'].set(value=peak.amplitude)

https://github.com/h2020charisma/ramanchada2/blob/cea31b41dc595842760ba5ee21a2abf577461402/src/ramanchada2/spectrum/peaks/fit_peaks.py#L62C13-L63C73

**Reply by evg-mar** [reply]
@vedina , @nicocopez , @georgievgeorgi 
After the above fix, we get the following result with the Nico's datasets.
![Figure_1](https://github.com/h2020charisma/ramanchada2/assets/10574551/ad15d510-546a-4ba5-b10a-4c2d5eff6328)

**Reply by georgievgeorgi** [reply]
@evg-mar, your changes look reasonable. Just keep in mind that sometimes `amplitude` means the height of the peak and sometimes means the integral.
In the last picture the fit looks good. I suspect that the x-axis of the orange line is quite rough. This is why it does not follow the blue cure

**Reply by nicocopez** [reply]
Was this change implemented? I just git pulled again and keep on encountering the same problem. Otherwise could you please provide the code that you used to get the above figure, to track what I am doing uncorrectly?

**Comment by vedina** [comment]
we should somehow try to find the reason rather than fixing things for particular dataset.  
according to previous explanation by @georgievgeorgi , amplitude in lmfit is the integral of the peak, and the height is the height.  
we need to check with the Pearson4 definition.   I do use Pearson4 in the calibration protocol and it works fine. 
So perhaps @nicocopez need just to use different parameters for fitting, rather than us changing the code.  
@nicocopez there should be kwargs options for the find_peak_multipeak

**Reply by nicocopez** [reply]
If it works fine for you, then I must be doing something wrong.
The find_peak_multipeak I use it for the peak_candidates, which is the same for all fittings, so I would think that this is correct. However, for the fit_peak_multimodel, where I specify the profile, maybe I can specify some other parameters. But I don't know how can I input there some new kwargs there?

**Comment by georgievgeorgi** [comment]
When i was working, the lmfit implementation was not accepted by the domain experts, This is why we have another implementation. Have a look here #63

**Comment by lellid** [comment]
Hi Nico, do you want Pearson4 to fit multiple peaks at once? As I mentioned, this is difficult, because we can not define parameter limits.

**Reply by nicocopez** [reply]
Not for this case, I am just fitting one silicon peak.

**Reply by evg-mar** [reply]
I have fixed it a few weeks ago. I will make a few more tests and let you know.

**Reply by evg-mar** [reply]
(@nicocopez ) As explained here, 
https://github.com/h2020charisma/ramanchada2/discussions/106#discussioncomment-8788100
I tested with several examples, and your example, Nico and it worked. But I don't know if this should be the right fix. I did it as I saw the analogy with the other methods (Pearson7, Moffat, Voigt), where 
fit_params[f'p{peak_i}_amplitude'] is set everywhere but only in Pearson4 it was fit_params[f'p{peak_i}_height']...

I don't consider myself a spectroscopy expert to say that 100% this is the right fix.

The change is not merged but you can test in in my branch 'fix_person4).

https://github.com/h2020charisma/ramanchada2/tree/fix_pearson4

**Comment by nicocopez** [comment]
Good morning!
If I understood it correctly from @vedina some changes have been done that are working for this example. But they are not yet availables in ramanchada2, correct? When will they be aproximately? Just to know. I need to use this function, so I can wait a bit more if it is gonna be usable, but if not I should be searching for alternatives. Thanks for the support!

**Reply by evg-mar** [reply]
Nico, the idea was you to confirm that you are Ok with the changes before we merge into the official release. I have posted above the repo above. You could test with more examples if you have.

**Reply by vedina** [reply]
@nicocopez the changes are now in the main branch . we have also added tests with synthetic spectra as well as with your example above
https://github.com/h2020charisma/ramanchada2/blob/main/tests/peak/pearson4_test.py

However,  my experience is the perason4 fitting is sensitive to the resolution , with small number of points it will not work (one can play with the number of bins of synthetic spectra in the test)

**Comment by nicocopez** [comment]
Ok, I think it is definitively me who is doing something wrong. I tried the test you send and it is not working on my laptop. Can you tell what am I missing? Maybe we can discuss it later on the meeting.

![imagen](https://github.com/h2020charisma/ramanchada2/assets/47475896/6d58ea09-2d9b-4869-a7cc-12c7f21222d6)

**Comment by vedina** [comment]
how did you load ramanchada2 ? You should load it from the main branch,  not from pip (we have not released it yet  in pip )
